[Fleet] Cascade agent policy's namespace to package policies

criamico opened the pull request:
## Summary

Cascade agent policy's namespace to package policies. This change introduces a new behavior for package policies namespace, mirroring the way outputs work:
- Allow leaving namespace blank when creating/updating a package policy. Agent policies namespace keeps being mandatory.
- If a package policy has a namespace defined, that's used to compile the inputs in the full agent policy.
- If a package policy has blank namespace, use the agent policy one.
- Namespace "default" is no longer used as default for package policy namespace across the code
- All the checks for the namespace are now moved to full agent policy handler. 

## UI
The only change in the UI is that the namespace in the policy creation can now be left blank:
![Screenshot 2024-01-16 at 16 55 07](https://github.com/elastic/kibana/assets/16084106/ed1be524-c4e6-49ac-8b53-39b07240d204)
I think that the text is still relevant, as nothing has changed from that point of view.

## Testing
### Creation - Inherit namespace
- Create agent policy with namespace `agentpolicyspace`
- Create package policy with blank namespace
- Generate full agent policy; package policy should have namespace `agentpolicyspace`

### Creation - Package policy custom namespace
- Create agent policy with namespace `agentpolicyspace`
- Create package policy with namespace `packagepolicyspace`
- Generate full agent policy; package policy should have namespace `packagepolicyspace`

### Update - Change default namespace
- Create agent policy with namespace  `agentpolicyspace`
- Create package policy with blank namespace
- Generate full agent policy; Package policy should have namespace `agentpolicyspace`
- Change agent policy namespace to `newspace`
- Package policy should have namespace `newspace` as well

### Update - Don't override custom namespace
- Create agent policy with namespace `agentpolicyspace`
- Create package policy with namespace `packagepolicyspace`
- Change agent policy namespace to `newspace`

### Check full agent policy
- namespace should be correct; check inputs for the package policies and output permissions as well

### Checklist
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

criamico commented:
/ci

criamico commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

jlind23 commented:
> ⚠️ This is marked as a breaking change, I'll add it to the release notes for 8.13.0. The API structure doesn't change (only difference is that namespace is now optional for package policy creation and update), but the behaviour is now different.

@lucabelluccini do you know how much of an impact this would cause

@nimarezainia Are you ok with this approach?

kpollich commented:
I tested this and it works well as far as I can tell, however I'm still not sure that the way we're handling preexisting policies is completely expected, e.g. 

1. Create a local environment with a few agent/integration policies off of `main`, leaving the agent policy namespace value as `default`
2. Observe that your agent policies and integration policies all have the `default` namespace defined explicitly via their respective edit pages
3. Check out this PR
4. Change the "Default namespace" value on one of your agent policies to `my_new_namespace`
5. Observe that your existing integration policies do not have their namespaces changed from `default`
6. Create a new policy, observe that it can now have a blank namespace and will properly "inherit" the new namespace value of `my_new_namespace` when its own namespace is not explicitly defined

This workflow definitely avoids breaking changes for existing customers, but I wonder if it feels unexpected at all. To me, it seems like if I have a default namespace of `default`, and _all_ my integration policies are also set to `default` and I've never explicitly set that value as a user - I would expect changing the agent policy default namespace to change all of my integration policy namespaces.

Do you think that's a fair expectation or am I off base? Does a one-time migration make sense for this case? We could blank out all integration policies where `integration_policy.namespace === agent_policy.namespace` for instance to move customers onto the new default behavior without interruption.

jlind23 commented:
I really feel like we should avoid breaking changes at all cost so this behavior makes sense to me.

nchaulet commented:
> Do you think that's a fair expectation or am I off base? Does a one-time migration make sense for this case? We could blank out all integration policies where integration_policy.namespace === agent_policy.namespace for instance to move customers onto the new default behavior without interruption.

In addition to add some migration complexity, I am wondering if it could cause the opposite effect as it change the behavior we currently have where updating agent policy do not exists in place package policy if namespace is defined, are we better to document this with a release not? if you want the agent policy namespace be the source of truth, you should not define a namespace in package policy...

nchaulet commented on x-pack/plugins/fleet/server/routes/package_policy/handlers.ts:
```diff
@@ -371,7 +371,7 @@ export const updatePackagePolicyHandler: FleetRequestHandler<
         name: body.name ?? packagePolicy.name,
         description: body.description ?? packagePolicy.description,
-        namespace: body.namespace ?? packagePolicy.namespace,
+        namespace: body?.namespace ?? packagePolicy?.namespace,
```
nit: does the `?` is needed here could `body` be undefined?

criamico replied:
Good catch, fixing it!

nchaulet commented on x-pack/plugins/fleet/server/types/models/agent_policy.ts:
```diff
@@ -8,8 +8,19 @@
+export const AgentPolicyNamespaceSchema = schema.string({
+  minLength: 1,
+  validate: (value) => {
+    const namespaceValidation = isValidNamespace(value || '');
```
👍 nice addition

nimarezainia commented:
> > ⚠️ This is marked as a breaking change, I'll add it to the release notes for 8.13.0. The API structure doesn't change (only difference is that namespace is now optional for package policy creation and update), but the behaviour is now different.
> 
> @lucabelluccini do you know how much of an impact this would cause?
> 
> @nimarezainia Are you ok with this approach?
> 
> @criamico @kpollich shall we make this the default behaviour only for newly installed integration instead?

Thanks for bringing this to my attention.
I may be misunderstanding this,  I've read the original issue also and consider this more of a bug than a breaking change. So correct me if I am wrong. 

I was *always* under the assumption that when the namespace on the agent policy is changed, that is propagated to all integration in that policy. if the user needs to get a more granular namespace defined then they can go to the integration and modify the namespace which will only apply to that input (package). Seems like this workflow is now implemented with ths PR.

I see the Agent Policy at the root node and everything under that inherits the namespace. Integrations are lower and inherit first but can be over written by the user.

We do need this behavior for the work on Fleet Space awareness.

criamico commented:
> I was always under the assumption that when the namespace on the agent policy is changed, that is propagated to all integration in that policy. if the user needs to get a more granular namespace defined then they can go to the integration and modify the namespace which will only apply to that input (package). Seems like this workflow is now implemented with ths PR.
@nimarezainia Thanks, I changed the release note to "fix".

criamico commented:
> From what I've understood, it seems we allowed to define a default namespace at policy level and to specify a namespace at package/integration level and until this fix is merged, it was not properly taken into account (?).

@lb yes, currently the namespace at agent policy level doesn't really propagate at package policy level. The current behavior has been like this for long time but it's not what users would expect, that's why we were debating if this is a bug fix or not. Imo from users perspective it is, it's much more intuitive to inherit the namespace from agent policy level to the package policies and define an override only if the user wants it.

>Let's preview the inherited namespace in the UI (when editing the package policy)

I need to check how to do it, currently I had to remove the inherited namespace from the form as it needs to be left blank. Maybe I can add it as a label outside of the form or something like that.

> Ensure default is still the agent policy namespace default



✅  This hasn't changed, I only modified the package policies one (and left 'default' as a fallback value [here](https://github.com/elastic/kibana/pull/174776/files#diff-ee6c1fe752205768ab54e6107a869041bcbb6130a0c982fa169bf5aba570a30eR56))

> For existing package policies created in versions prior this PR, the default value was already present in the SO.
Those saved objects will be kept as is?


Yes, we are not doing any migration in this PR (see [this comment](https://github.com/elastic/kibana/pull/174776#issuecomment-1894258841)). @kpollich has [an example](https://github.com/elastic/kibana/pull/174776#issuecomment-1894231257) of what would happen to an existing policy when this PR is checked it doesn't affect the existing SOs.

juliaElastic commented on x-pack/plugins/fleet/server/types/models/preconfiguration.ts:
```diff
@@ -130,7 +130,7 @@ export const PreconfiguredFleetProxiesSchema = schema.arrayOf(
   schema.object({
     ...AgentPolicyBaseSchema,
-    namespace: schema.maybe(NamespaceSchema),
+    namespace: schema.maybe(AgentPolicyNamespaceSchema),
```
shouldn't the agent policy namespace be required here?

nchaulet replied:
I think this will introduce a breaking change, but we should probably have a default value no?

criamico replied:
I'm not sure why there was a `schema.maybe` there. I think you right, I didn't notice it, I was basically enforcing it from the schema itself (has `lenght: 1` as requirement). I'll fix it

criamico replied:
@nchaulet that's true 🤦‍♀️ We could leave as it is, the AgentPolicySchema actually enforces it already:
https://github.com/elastic/kibana/pull/174776/files#diff-7c570055afd7a69dcc80854cb7559a0735b7580eab6ec1254e7344bea549a274R15-R23

To clarify, this is not doing any change on the current functionality, it's just separating the schema for agent policy namespace (that requires to have namespace) and package policy namespace (that can be optional). So I wouldn't do any further changes here.

juliaElastic replied:
wouldn't it be enough to use the same `NamespaceSchema` with a `schema.maybe` wrapped around when using it in package policy?

criamico replied:
> wouldn't it be enough to use the same NamespaceSchema with a schema.maybe wrapped around when using it in package policy?

I had it like that and without the `minLength: 1` the agent policy schema allowed the field to be empty, that's why I redefined it. I don't think it is enough.

criamico commented:
I pushed a [new commit](https://github.com/elastic/kibana/pull/174776/commits/c4c54f97062f644b9e9567741ad1e5d4ed266a62) with a couple of improvements for the UI, following on @lucabelluccini feedback.

1. The integration policy form will now show the inherited namespace as a placeholder, to make clear that it will be the agent policy one if left blank:

![Screenshot 2024-01-17 at 15 39 35](https://github.com/elastic/kibana/assets/16084106/82d8ebcc-1c21-493c-ac73-7fdbf7769ecd)

2. In the integration policies table I added a different badge with a tooltip that clarifies that the namespace is defined in the agent policy, instead of showing nothing:
![Screenshot 2024-01-17 at 15 32 51](https://github.com/elastic/kibana/assets/16084106/696ca46b-f440-4884-9749-7f5957260233)

Let me know if any more changes are needed in the UI to resurface the new behavior, or if you have any feedback on the copy.

criamico commented:
@elasticmachine merge upstream

kilfoyle commented:
> @/kilfoyle We would like to make sure that this change is highlighted in the release notes for 8.13.0. Should we add it to the docs as well?

Certainly! For the docs, I've opened up a PR to add a small section explaining the cascading behaviour: https://github.com/elastic/ingest-docs/pull/822 Please let me know if I got things right.

For the release notes, the change fragment should automatically go into the Kibana Release Notes, and I copy the Fleet contents from the Kibana Release Notes into the Fleet and Agent Release Notes, so the change (a bug fix I think) will appear in both books.

criamico commented:
@elasticmachine merge upstream

criamico commented:
@elasticmachine merge upstream

juliaElastic commented:
/ci

nchaulet reviewed (approved):
LGTM seems to work as expected 🚀

criamico commented:
@elasticmachine merge upstream

kilfoyle commented on x-pack/plugins/fleet/common/openapi/bundled.json:
```diff
@@ -7671,8 +7671,8 @@
             "type": "string",
-            "description": "namespace by default \"default\"",
-            "example": "default"
+            "description": "package policy namespace. It can be left blank to inherit agent policy's one",
```
```suggestion
            "description": "The package policy namespace. Leave blank to inherit the agent policy's namespace.",
```
I'd rephrase like this (and of course the same for the other descriptions).

criamico replied:
Fixed!

criamico commented:
@elasticmachine merge upstream

kilfoyle reviewed (approved):
LGTM for the text. 👍
